### PR TITLE
Cleanup expired emails

### LINKED ISSUE #32: RFC: Automatically destroy expired emails
### Proposal

1: Automatically really destroy deleted emails
2: Automatically really destroy ignored (old) emails

### Why

Mailgun only holds emails for 3 days, after that, the emails get destroyed. The only emails that are not automatically forwarded in our email system are emails to an alias for a (semi-)closed mail alias.

When a `StoredMail` is not moderated within 3 days, it expires and should not live in our database anymore.

When a `StoredMail` is accepted or rejected, it is soft-deleted.

The purpose of soft-deletion is recovery and debugging issues with this feature critical to our association. However, there are privacy-reasons for not keeping them forever, and after 3 days they are not relevant anymore anyway.

### How

My suggestion is to execute a CleanupExpiredStoredMailsJob every day, that:

1. `really_destroy!`s 2-week-old soft-deleted emails
1. `destroy` 4 days old ignored emails

That way, we reduce the amount of irrelevant (for us) but potential privacy-compromising information in our database, but still allow a grace-period for debugging issues.

I estimate that a 2-week window before really destroying emails is enough for a member to 1) spot a potential email loss and 2) tell us about it – whereas 1 week will be too short.

### Extra

I think we also need to inform a user if an email expired. The moderators would get an email like `"The email with subject #{something} has expired".`

### PR DESCRIPTION
**Pull request by cmitz**
### Checklist
- [nvt] merged database migrations into 1 database migration
- [nvt] tested database migrations from `origin/staging` (`git checkout staging ; git pull ; bundle exec rails db:reset ; git checkout BRANCH ; bundle exec rails db:migrate`)

### Summary
Adds a job that cleans up old stored emails from the database

### Other

Implements and fixes #32

### CODE REVIEW

File: app/jobs/cleanup_expired_stored_mails_job.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    # Delete all expired (ignored) emails after the 3-day threshold imposed by Mailgun
+    expired_mails = StoredMail.where('created_at <= ?', 4.days.ago)
+    expired_mails_count = expired_mails.count
+    expired_mails.each(&:destroy) if expired_mails.any?
+
+    inform_slack(deleted_mails_count, expired_mails_count)
+  end
+
+  def inform_slack(deleted_count, expired_count)
+    if deleted_count.positive? && expired_count.positive?
```
Comment (tcoenraad) [inline app/jobs/cleanup_expired_stored_mails_job.rb]: If you early return this, it becomes easier to read the way smaller `else` branch.

File: app/jobs/cleanup_expired_stored_mails_job.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class CleanupExpiredStoredMailsJob < ApplicationJob
+  queue_as :default
+
+  def perform
+    # Really destroy already moderated emails older than two weeks
```
Comment (Matthijsy) [inline app/jobs/cleanup_expired_stored_mails_job.rb]: Comment doesn't match the 4 weeks in the next line, I think 4 weeks is good

File: app/jobs/cleanup_expired_stored_mails_job.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  def inform_slack(deleted_count, expired_count)
+    if deleted_count.positive? && expired_count.positive?
+      SlackMessageJob.perform_later(
+        "Email cleanup finished. #{deleted_count} expired emails really destroyed; "\
+        "#{expired_count} ignored (not-moderated) emails soft-deleted",
+        channel: '#monitoring'
+      )
+    else
+      SlackMessageJob.perform_later('No email cleanup needed', channel: '#monitoring')
```
Comment (Matthijsy) [inline app/jobs/cleanup_expired_stored_mails_job.rb]: Do we need a no cleanup message? Not that interessting right?

File: spec/jobs/cleanup_expired_stored_mails_job_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+      it do
+        expect(slack_message_job_class).to have_received(:perform_later)
+          .with('No email cleanup needed', channel: '#monitoring')
+      end
+    end
+
+    context 'when with some deleted emails and expired emails' do
+      let(:expired_email) { FactoryBot.create(:stored_mail, :expired) }
+      let(:deleted_email) { FactoryBot.create(:stored_mail, :deleted) }
```
Comment (Matthijsy) [inline spec/jobs/cleanup_expired_stored_mails_job_spec.rb]: you can use create_list right?

File: spec/jobs/cleanup_expired_stored_mails_job_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        job.perform
+      end
+
+      it do
+        expect(slack_message_job_class).to have_received(:perform_later).with(
+          'Email cleanup finished. 2 expired emails really destroyed; '\
+          '1 ignored (not-moderated) emails soft-deleted',
+          channel: '#monitoring'
+        )
+      end
```
Comment (Matthijsy) [inline spec/jobs/cleanup_expired_stored_mails_job_spec.rb]: Also check that the stored emails are really destored or soft destroyed?

File: app/jobs/cleanup_expired_stored_mails_job.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class CleanupExpiredStoredMailsJob < ApplicationJob
+  queue_as :default
+
+  def perform
+    # Really destroy already moderated emails older than two weeks
+    deleted_mails = StoredMail.only_deleted.where('created_at <= ?', 4.weeks.ago)
+    deleted_mails_count = deleted_mails.count
+    deleted_mails.each(&:really_destroy!) if deleted_mails_count.positive?
```
[the remark this answers is not in the document]
Comment (cmitz) [reply app/jobs/cleanup_expired_stored_mails_job.rb]: Yes, you're right. I first tried to do `#really_destroy!` on a collection, there the if clause was neccessary. But that destroy method doesn't exist on collections so I had to do `#each`. The if clause was debt

File: app/jobs/cleanup_expired_stored_mails_job.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    expired_mails.each(&:destroy)
+
+    inform_slack(deleted_mails_count, expired_mails_count)
+  end
+
+  def inform_slack(deleted_count, expired_count)
+    return unless deleted_count.positive? || expired_count.positive?
+
+    SlackMessageJob.perform_later(
+      "Email cleanup finished. #{deleted_count} expired #{'emails'.pluralize(deleted_count)} "\
```
Comment (Matthijsy) [inline app/jobs/cleanup_expired_stored_mails_job.rb]: shouldn't that be 'email'.pluralize(deleted_count)? (without the s)

**Pull request merged**